kani-compiler/src/kani_middle/stubbing/transform.rs

```diff
@@ -23,18 +26,57 @@ pub fn get_stub(tcx: TyCtxt, def_id: DefId) -> Option<DefId> {
 
 /// Returns the new body of a function/method if it has been stubbed out;
 /// otherwise, returns the old body.
-pub fn transform<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    def_id: DefId,
-    old_body: &'tcx Body<'tcx>,
-) -> &'tcx Body<'tcx> {
+pub fn transform<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, old_body: &'tcx Body<'tcx>) -> Body<'tcx> {
     if let Some(replacement) = get_stub(tcx, def_id) {
         let new_body = tcx.optimized_mir(replacement).clone();
         if check_compatibility(tcx, def_id, old_body, replacement, &new_body) {
-            return tcx.arena.alloc(new_body);
+            return new_body;
+        }
+    }
+    old_body.clone()
+}
+
+/// Tranverse `body` searching for calls to foreing functions and, whevever there is
+/// a stub available, replace the call to the foreign function with a call
+/// to its correspondent stub. This happens as a separate step because there is no
+/// body available to foreign functions at this stage.
+pub fn transform_foreign_functions<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let mut visitor = if let Some(stub_map) = get_stub_mapping(tcx) {
+        ForeignFunctionTransformer { tcx, body: body.clone(), stub_map }
+    } else {
+        ForeignFunctionTransformer { tcx, body: body.clone(), stub_map: HashMap::default() }
+    };
+    visitor.visit_body(body);
+}
+
+struct ForeignFunctionTransformer<'tcx> {
+    /// The compiler context.
+    tcx: TyCtxt<'tcx>,
+    /// Body of the callee function. Kani searches here for foreign functions.
+    body: Body<'tcx>,
```

Comment:
As far as I can tell you're only using this for `operand.ty` so it ould be enough if you cloned and stored just `LocalDecls` here.

Reply:
Why would that be better? It looks clean and we might want to reuse this for future transformations and having access to the body might still be useful. Is there any huge drawback I might be missing?

Reply:
Well you're creating a deep copy of quite a large structure without specific reason to do so.